### PR TITLE
Better goal positions in levels 7 and 8.

### PR DESCRIPTION
**Pull request by islemaster**
Adjusting some offsets and making things a little easier:

Here's level 7 before:
screenshot from 2015-10-22 15 55 35

and after (note that the characters are horizontally centered):
screenshot from 2015-10-22 15 57 13

And here's level 8 before:
screenshot from 2015-10-22 15 57 52

and after (fewer turns to reach the goals):
screenshot from 2015-10-22 16 02 05

In addition, the timeout on level 8 has gone from 20 to 30 seconds.

### CODE REVIEW

File: apps/src/studio/studio.js [hunk 1]
```diff
@@ -347,10 +347,9 @@ var drawMap = function () {
     }
   }
 
+  var goalOverride = utils.valueOr(level.goalOverride, {});
   var numFrames = 1;
-  if (level.goalOverride && 
-      level.goalOverride.goalAnimation && 
-      skin.animatedGoalFrames) {
+  if (goalOverride.goalAnimation && skin.animatedGoalFrames) {
```
Comment (islemaster) [inline apps/src/studio/studio.js:352]: Note: This should be a nonfunctional change, I've just extracted the null-check for goalOverride throughout `drawMap`.

File: apps/src/studio/studio.js [hunk 2]
```diff
@@ -373,15 +369,9 @@ var drawMap = function () {
...
-      width = numFrames * 
-        ((level.goalOverride && level.goalOverride.imageWidth) || Studio.MARKER_WIDTH);
-
-      height = 
-        (level.goalOverride && level.goalOverride.imageHeight) || Studio.MARKER_HEIGHT;
-
       var spriteFinishMarker = document.createElementNS(SVG_NS, 'image');
       spriteFinishMarker.setAttribute('id', 'spriteFinish' + i);
-      spriteFinishMarker.setAttribute('width', width);
+      spriteFinishMarker.setAttribute('width', numFrames * width);
```
Comment (islemaster) [inline apps/src/studio/studio.js:374]: Removing a bunch of redundant calculations.

File: apps/src/studio/studio.js [hunk 3]
```diff
@@ -1894,43 +1884,56 @@ Studio.reset = function(first) {
   for (i = 0; i < Studio.spriteCount; i++) {
     Studio.displaySprite(i);
   }
+  this.resetGoalSprites();
```
Comment (islemaster) [inline apps/src/studio/studio.js:1887]: Diff gets weird here as I extracted a large chunk of `reset()` into `resetGoalSprites()`.

File: apps/src/studio/studio.js [hunk 4]
```diff
@@ -1894,43 +1884,56 @@ Studio.reset = function(first) {
...
+};
+
+/**
+ * Move all goal sprites to their original positions, and reset their completion
+ * state, both visual and logical.
+ */
+Studio.resetGoalSprites = function () {
   Studio.touchAllGoalsEventFired = false;
-  for (i = 0; i < Studio.spriteGoals_.length; i++) {
+  for (var i = 0; i < Studio.spriteGoals_.length; i++) {
```
Comment (islemaster) [inline apps/src/studio/studio.js:1903]: Code smell: Any time you see yourself reusing a declaration of `i` for many loops, your function is too long!

File: apps/src/studio/studio.js [hunk 5]
```diff
@@ -1894,43 +1884,56 @@ Studio.reset = function(first) {
...
     Studio.spriteGoals_[i].startFadeTime = null;
 
     // Move the finish icons into position.
-    var offsetX = skin.goalRenderOffsetX || 0;
-    var offsetY = skin.goalRenderOffsetY || 0;
+    var goalOverride = utils.valueOr(level.goalOverride, {});
+    var offsetX = utils.valueOr(goalOverride.goalRenderOffsetX,
+        utils.valueOr(skin.goalRenderOffsetX, 0));
+    var offsetY = utils.valueOr(goalOverride.goalRenderOffsetY,
+        utils.valueOr(skin.goalRenderOffsetY, 0));
```
Comment (islemaster) [inline apps/src/studio/studio.js:1913]: Here's actual new functionality: The ability to set a goalRenderOffset on a per-level basis.  Rare, but useful for special cases like level 7.

File: apps/src/studio/studio.js [hunk 6]
```diff
@@ -1894,43 +1884,56 @@ Studio.reset = function(first) {
...
     var spriteFinishIcon = document.getElementById('spriteFinish' + i);
     spriteFinishIcon.setAttribute('x', Studio.spriteGoals_[i].x + offsetX);
     spriteFinishIcon.setAttribute('y', Studio.spriteGoals_[i].y + offsetY);
     spriteFinishIcon.setAttributeNS('http://www.w3.org/1999/xlink',
-      'xlink:href', goalAsset);
+        'xlink:href', Studio.getGoalAssetFromSkin());
```
Comment (islemaster) [inline apps/src/studio/studio.js:1918]: `goalAsset` was originally calculated 20 lines ago and only used once here; now extracted into a helper method below.

**Comment by breville** [comment]
lgtm.  worth testing with some legacy levels that used goal overrides to some degree but not to the degree we do for this tutorial.  as well as the usual legacy candidates like infinity and playlab scripts.

**Comment by islemaster** [comment]
The only level outside of hoc2015 that currently uses `goalOverride` is s/playlab/stage/1/puzzle/7 which looks exactly the same before and after this change.  That's... a weird look, but it is the same.

**Pull request merged**